src/engine/QueryPlanner.cpp

```diff
@@ -2516,6 +2619,40 @@ void QueryPlanner::QueryGraph::setupGraph(
             }
           }
         }
+
+        // Add additional edges to the graph representing the connections
+        // between variables given by joins substituting cartesian product +
+        // filter.
+        for (auto& [filter, substitute] : filtersAndOptionalSubstitutes) {
+          if (!substitute.has_value()) {
+            // This filter cannot be substituted: add no edges.
+            continue;
+          }
+          std::vector<Variable> varsToBeConnected;
+          for (auto var : filter.expression_.containedVariables()) {
+            if (varToNode.contains(*var)) {
+              varsToBeConnected.push_back(*var);
+            }
```

Comment:
1. If you get stable pointers, then you can also store pointers in the vector.
2. if you use `absl::InlinedVector` for the first few pointers, then you can easily avoid any memory allocations.